### PR TITLE
Feature/tcvp 1199 email service

### PR DESCRIPTION
**Pull request by elainecwong**
# Description

This PR includes the following proposed change(s):

TCVP-1199
Added the SendEmailConsumer, and all required objects EmailSenderService and Smtp connection client.
Added Xunit tests for the new consumer and service

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Refactoring / Documentation
- [ ] Version change

if your change is a breaking change, please add `breaking change` label to this PR

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes.

## Does the change impact or break the Docker build?

- [ ] Yes
- [x] No

If Yes: Has Docker been updated accordingly?

- [ ] Yes
- [x] No

## Checklist:

- [ ] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: src/backend/TrafficCourts/Messaging/MessageContracts/SendEmail.cs [hunk 1]
```diff
@@ -0,0 +1,19 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace TrafficCourts.Messaging.MessageContracts
+{
+    public interface SendEmail: IMessage
```
[the remark this answers is not in the document]
Comment (elainecwong) [reply src/backend/TrafficCourts/Messaging/MessageContracts/SendEmail.cs]: done

File: src/backend/TrafficCourts/Workflow.Service/Services/EmailSenderService.cs [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+
+            }
+        }
+
+        // Validation to check if the passed-in e-mail is in a valid format.
+        private static bool IsValidEmail(string emailAddress)
+        {
+            try
+            {
+                System.Net.Mail.MailAddress m = new System.Net.Mail.MailAddress(emailAddress);
```
[the remark this answers is not in the document]
Comment (elainecwong) [reply src/backend/TrafficCourts/Workflow.Service/Services/EmailSenderService.cs]: Applied change, as per suggested in IsEmailAllowed method

File: src/backend/TrafficCourts/Workflow.Service/Services/EmailSenderService.cs [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+            catch (FormatException)
+            {
+                return false;
+            }
+        }
+
+        // Go through list of allowed emails, and make sure it matches
+        private bool IsEmailAllowed(string emailAddress)
+        {
+            if(_emailConfiguration.AllowList != null)
```
[the remark this answers is not in the document]
Comment (elainecwong) [reply src/backend/TrafficCourts/Workflow.Service/Services/EmailSenderService.cs]: Replaced all null checks with is null/is not null

File: src/backend/TrafficCourts/Workflow.Service/Services/EmailSenderService.cs [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+            {
+                return false;
+            }
+            catch (FormatException)
+            {
+                return false;
+            }
+        }
+
+        // Go through list of allowed emails, and make sure it matches
```
[the remark this answers is not in the document]
Comment (elainecwong) [reply src/backend/TrafficCourts/Workflow.Service/Services/EmailSenderService.cs]: Added XML comments to various methods/functions

File: src/backend/TrafficCourts/Workflow.Service/Services/EmailSenderService.cs [hunk 5]
```diff
@@ -0,0 +1,216 @@
...
+            }
+            return false;
+        }
+    }
+
+
+    [Serializable]
+    internal class EmailSendFailedException : Exception
+    {
+        public EmailSendFailedException() { }
```
[the remark this answers is not in the document]
Comment (elainecwong) [reply src/backend/TrafficCourts/Workflow.Service/Services/EmailSenderService.cs]: Cleaned-up unused constructors

File: src/backend/TrafficCourts/Workflow.Service/Services/EmailSenderService.cs [hunk 6]
```diff
@@ -0,0 +1,216 @@
...
+                else if (!String.IsNullOrEmpty(emailMessage.HtmlContent))
+                {
+                    email.Body = new TextPart(TextFormat.Html) { Text = emailMessage.HtmlContent };
+                }
+
+                // Should do a check to see if mandatory fields exist.
+                if(String.IsNullOrEmpty(emailMessage.Subject) || email.Body == null /*||
+                    (email.To.Count == 0 && email.Cc.Count == 0 && email.Bcc.Count == 0)*/)
+                {
+                    throw new ArgumentNullException();
```
[the remark this answers is not in the document]
Comment (elainecwong) [reply src/backend/TrafficCourts/Workflow.Service/Services/EmailSenderService.cs]: Added custom exception, with error message clarifying issue.

File: src/backend/TrafficCourts/Workflow.Service/Services/EmailSenderService.cs [hunk 7]
```diff
@@ -0,0 +1,216 @@
...
+            _smptClientFactory = stmpClientFactory;
+        }
+
+        public async Task SendEmailAsync(EmailMessage emailMessage)
+        {
+            try
+            {
+                // create email message
+                var email = new MimeMessage();
+                email.From.Add(MailboxAddress.Parse(emailMessage.From??_emailConfiguration.Sender));
```
[the remark this answers is not in the document]
Comment (elainecwong) [reply src/backend/TrafficCourts/Workflow.Service/Services/EmailSenderService.cs]: fixed formatting

File: src/backend/TrafficCourts/Workflow.Service/Services/EmailSenderService.cs [hunk 8]
```diff
@@ -0,0 +1,216 @@
...
+
+                if(emailMessage.To != null)
```
[the remark this answers is not in the document]
Comment (elainecwong) [reply src/backend/TrafficCourts/Workflow.Service/Services/EmailSenderService.cs]: Applied refactored changes.

File: src/backend/TrafficCourts/Workflow.Service/Services/EmailSenderService.cs [hunk 9]
```diff
@@ -0,0 +1,216 @@
...
+        private readonly ISmtpClientFactory _smptClientFactory;
+
+        public EmailSenderService(ILogger<EmailSenderService> logger, IOptions<EmailConfiguration> emailConfiguration, ISmtpClientFactory stmpClientFactory)
+        {
+            _logger = logger;
+            _emailConfiguration = emailConfiguration.Value;
+            _smptClientFactory = stmpClientFactory;
+        }
+
+        public async Task SendEmailAsync(EmailMessage emailMessage)
```
[the remark this answers is not in the document]
Comment (elainecwong) [reply src/backend/TrafficCourts/Workflow.Service/Services/EmailSenderService.cs]: Updated to use consumer's cancellation token, and updated unit tests as required.

File: src/backend/TrafficCourts/Workflow.Service/Services/EmailSenderService.cs [hunk 10]
```diff
@@ -0,0 +1,216 @@
...
+            catch(SmtpConnectFailedException scfe)
+            {
+                // An error connecting to the the SMTP Server
+                _logger.LogError(scfe, "An error connecting to the the SMTP Server.");
+                throw new EmailSendFailedException($"An error connecting to the the SMTP Server", scfe);
+            }
+            catch (Exception ex)
+            {
+                _logger.LogError(ex, "General email send exception thrown.");
+                throw new EmailSendFailedException($"General email send exception thrown", ex);
```
[the remark this answers is not in the document]
Comment (elainecwong) [reply src/backend/TrafficCourts/Workflow.Service/Services/EmailSenderService.cs]: fixed

File: src/backend/TrafficCourts/Workflow.Service/Configuration/EmailConfiguration.cs [hunk 11]
```diff
@@ -0,0 +1,8 @@
+﻿namespace TrafficCourts.Workflow.Service.Configuration
+{
+    public class EmailConfiguration
+    {
+        public string? Sender { get; set; }
+        public String[]? AllowList { get; set; }
```
[the remark this answers is not in the document]
Comment (elainecwong) [reply src/backend/TrafficCourts/Workflow.Service/Configuration/EmailConfiguration.cs]: fixed

File: src/backend/TrafficCourts/Workflow.Service/Services/EmailSenderService.cs [hunk 12]
```diff
@@ -0,0 +1,216 @@
...
+                {
+
+                    Match m = Regex.Match(emailAddress, allowedEmail, RegexOptions.IgnoreCase);
+                    if (m.Success)
+                    {
+                        return true;
+                    }
+                }
+            }
+            return false;
```
[the remark this answers is not in the document]
Comment (elainecwong) [reply src/backend/TrafficCourts/Workflow.Service/Services/EmailSenderService.cs]: Bad logic.  Added new unit test, to make sure it is working properly.

File: src/backend/TrafficCourts/Workflow.Service/Models/EmailMessage.cs [hunk 13]
```diff
@@ -0,0 +1,28 @@
...
+    {
+        public string From { get; set; }
+        public string[] To { get; set; }
+        public string[] Cc { get; set; }
+        public string[] Bcc { get; set; }
+        public string Subject { get; set; }
+        public string PlainTextContent { get; set; }
+        public string HtmlContent { get; set; }
+
+        public override string ToString()
```
[the remark this answers is not in the document]
Comment (elainecwong) [reply src/backend/TrafficCourts/Workflow.Service/Models/EmailMessage.cs]: Override removed.  Leftover code from copy/paste of another model object.

File: src/backend/TrafficCourts/Workflow.Service/Configuration/EmailConfiguration.cs [hunk 14]
```diff
@@ -0,0 +1,8 @@
+﻿namespace TrafficCourts.Workflow.Service.Configuration
+{
+    public class EmailConfiguration
```
[the remark this answers is not in the document]
Comment (elainecwong) [reply src/backend/TrafficCourts/Workflow.Service/Configuration/EmailConfiguration.cs:6]: Added comments to the configuration classes.

File: src/backend/TrafficCourts/Workflow.Service/Services/EmailSenderService.cs [hunk 15]
```diff
@@ -0,0 +1,216 @@
...
+
+        // Go through list of allowed emails, and make sure it matches
+        private bool IsEmailAllowed(string emailAddress)
+        {
+            if(_emailConfiguration.AllowList != null)
+            {
+                foreach (var allowedEmail in _emailConfiguration.AllowList)
+                {
+
+                    Match m = Regex.Match(emailAddress, allowedEmail, RegexOptions.IgnoreCase);
```
[the remark this answers is not in the document]
Comment (elainecwong) [reply src/backend/TrafficCourts/Workflow.Service/Services/EmailSenderService.cs]: Applied changes

**Review by pbolduc — APPROVED** [review]
There are some other minor changes I would like to see, however, it is good enough for now.

**Pull request merged**